#### This is the assignment for the IBM Data Science Capstone week 3

Import panadas and create beautiful soup

In [107]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
source=requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
soup = BeautifulSoup(source,'html.parser')
soup.title
report=pd.DataFrame(columns=['Postcode','Borough','Neighborhood'])

In [108]:
df = pd.DataFrame([[1,2,3]],columns=['Postcode','Borough','Neighborhood']) # I just wanted to see what the frame looks like with 1/2/3 digits, I'm new to this
for i in range(0,289):
    s = pd.Series([soup.findAll('td')[i*3+0].text,soup.findAll('td')[i*3+1].text,soup.findAll('td')[i*3+2].text],index=['Postcode','Borough','Neighborhood'])
    df=df.append(s,ignore_index=True)   
  

In [109]:
df['Neighborhood']=df['Neighborhood'].str.replace("\n", "") # remove some characters that we don't want

In [110]:
df.head() # 

Postcode       Borough      Neighborhood
0        1             2               NaN
1      M1A  Not assigned      Not assigned
2      M2A  Not assigned      Not assigned
3      M3A    North York         Parkwoods
4      M4A    North York  Victoria Village

In [111]:
df=df[df.Borough != 'Not assigned']
df=df[df.Borough != 2] # remove my first line

In [112]:
df['Neighborhood']=df['Neighborhood'].replace('Not assigned',df['Borough'])

In [113]:
df['Postcode'] = df['Postcode'].astype('str') 

In [114]:
df2=df.groupby(['Postcode','Borough'], as_index=False).agg(','.join)

In [115]:
type(df2) # making sure this is still a dataframe

pandas.core.frame.DataFrame

### This is the first data frame that we're looking for 
1. Postal Code, Borough and Neighborhood
2. Remove unassigned
3. Group by Postcode/Borough and combine Neighborhoods
4. Shape will be below

In [127]:
df2.head(30)

Postcode      Borough                                       Neighborhood
0       M1B  Scarborough                                      Rouge,Malvern
1       M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2       M1E  Scarborough                    Guildwood,Morningside,West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7       M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8       M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9       M1N  Scarborough                         Birch Cliff,Cliffside West
10      M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11      M1R  Scarborough                                   Maryvale,Wexford
12      M1S  Scarborough                                          Agincourt
13      M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
14      M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...
15      M1W  Scarborough                       L'Amoreaux West,Steeles West
16      M1X  Scarborough                                        Upper Rouge
17      M2H   North York                                  Hillcrest Village
18      M2J   North York                         Fairview,Henry Farm,Oriole
19      M2K   North York                                    Bayview Village
20      M2L   North York                            Silver Hills,York Mills
21      M2M   North York                             Newtonbrook,Willowdale
22      M2N   North York                                   Willowdale South
23      M2P   North York                                    York Mills West
24      M2R   North York                                    Willowdale West
25      M3A   North York                                          Parkwoods
26      M3B   North York                                    Don Mills North
27      M3C   North York                    Flemingdon Park,Don Mills South
28      M3H   North York      Bathurst Manor,Downsview North,Wilson Heights
29      M3J   North York                     Northwood Park,York University

In [130]:
df2.shape

(103, 3)

It looks like 103 rows

#### Add in the geospatial coordinates from the csv file

In [128]:
df3 = pd.read_csv('Geospatial_Coordinates2.csv') # read in the coordinates from the csv

In [129]:
df3.head() 

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [133]:
df4=df3.merge(df2,on='Postcode')

In [138]:
df4=df4[['Postcode','Borough','Neighborhood','Latitude','Longitude']] # change the order of the columns 

#### This is the second dataframe in the assignment question

In [125]:
df4.head(50)

Postcode   Latitude  Longitude          Borough  \
0       M1B  43.806686 -79.194353      Scarborough   
1       M1C  43.784535 -79.160497      Scarborough   
2       M1E  43.763573 -79.188711      Scarborough   
3       M1G  43.770992 -79.216917      Scarborough   
4       M1H  43.773136 -79.239476      Scarborough   
5       M1J  43.744734 -79.239476      Scarborough   
6       M1K  43.727929 -79.262029      Scarborough   
7       M1L  43.711112 -79.284577      Scarborough   
8       M1M  43.716316 -79.239476      Scarborough   
9       M1N  43.692657 -79.264848      Scarborough   
10      M1P  43.757410 -79.273304      Scarborough   
11      M1R  43.750072 -79.295849      Scarborough   
12      M1S  43.794200 -79.262029      Scarborough   
13      M1T  43.781638 -79.304302      Scarborough   
14      M1V  43.815252 -79.284577      Scarborough   
15      M1W  43.799525 -79.318389      Scarborough   
16      M1X  43.836125 -79.205636      Scarborough   
17      M2H  43.803762 -79.363452       North York   
18      M2J  43.778517 -79.346556       North York   
19      M2K  43.786947 -79.385975       North York   
20      M2L  43.757490 -79.374714       North York   
21      M2M  43.789053 -79.408493       North York   
22      M2N  43.770120 -79.408493       North York   
23      M2P  43.752758 -79.400049       North York   
24      M2R  43.782736 -79.442259       North York   
25      M3A  43.753259 -79.329656       North York   
26      M3B  43.745906 -79.352188       North York   
27      M3C  43.725900 -79.340923       North York   
28      M3H  43.754328 -79.442259       North York   
29      M3J  43.767980 -79.487262       North York   
30      M3K  43.737473 -79.464763       North York   
31      M3L  43.739015 -79.506944       North York   
32      M3M  43.728496 -79.495697       North York   
33      M3N  43.761631 -79.520999       North York   
34      M4A  43.725882 -79.315572       North York   
35      M4B  43.706397 -79.309937        East York   
36      M4C  43.695344 -79.318389        East York   
37      M4E  43.676357 -79.293031     East Toronto   
38      M4G  43.709060 -79.363452        East York   
39      M4H  43.705369 -79.349372        East York   
40      M4J  43.685347 -79.338106        East York   
41      M4K  43.679557 -79.352188     East Toronto   
42      M4L  43.668999 -79.315572     East Toronto   
43      M4M  43.659526 -79.340923     East Toronto   
44      M4N  43.728020 -79.388790  Central Toronto   
45      M4P  43.712751 -79.390197  Central Toronto   
46      M4R  43.715383 -79.405678  Central Toronto   
47      M4S  43.704324 -79.388790  Central Toronto   
48      M4T  43.689574 -79.383160  Central Toronto   
49      M4V  43.686412 -79.400049  Central Toronto   

                                         Neighborhood  
0                                       Rouge,Malvern  
1                Highland Creek,Rouge Hill,Port Union  
2                     Guildwood,Morningside,West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                                 Scarborough Village  
6           East Birchmount Park,Ionview,Kennedy Park  
7                       Clairlea,Golden Mile,Oakridge  
8       Cliffcrest,Cliffside,Scarborough Village West  
9                          Birch Cliff,Cliffside West  
10  Dorset Park,Scarborough Town Centre,Wexford He...  
11                                   Maryvale,Wexford  
12                                          Agincourt  
13              Clarks Corners,Sullivan,Tam O'Shanter  
14  Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                       L'Amoreaux West,Steeles West  
16                                        Upper Rouge  
17                                  Hillcrest Village  
18                         Fairview,Henry Farm,Oriole  
19                                    Bayview Village  
20                            Silver Hills,York Mills  
21              

#### Now let's take a look at the Toronto Maps
Import first

In [25]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [140]:
address = 'Toronto'  # following the NYC lab just to see what happens

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [141]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [142]:
toronto_data = df4[df4['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
toronto_data.head()

Postcode           Borough                Neighborhood   Latitude  Longitude
0      M4W  Downtown Toronto                    Rosedale  43.679563 -79.377529
1      M4X  Downtown Toronto  Cabbagetown,St. James Town  43.667967 -79.367675
2      M4Y  Downtown Toronto        Church and Wellesley  43.665860 -79.383160
3      M5A  Downtown Toronto    Harbourfront,Regent Park  43.654260 -79.360636
4      M5B  Downtown Toronto     Ryerson,Garden District  43.657162 -79.378937

#### Instead of using any borough with Toronto, I just used Downtown Toronto

Create the map

In [143]:
# create map of Manhattan using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [53]:
CLIENT_ID = 'WJSQRHGQU0GVDHLEMW2IOS23DBIOMACV2X5VBVCWMV11S1H3' # your Foursquare ID
CLIENT_SECRET = 'NB1WGADMKQ2NPNHBF51BZL3GBDGLSQ15Z1NDPCX25JIHQMOB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WJSQRHGQU0GVDHLEMW2IOS23DBIOMACV2X5VBVCWMV11S1H3
CLIENT_SECRET:NB1WGADMKQ2NPNHBF51BZL3GBDGLSQ15Z1NDPCX25JIHQMOB


In [54]:
toronto_data.loc[0, 'Neighborhood']

'Rosedale'

In [55]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


In [56]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius


In [57]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=WJSQRHGQU0GVDHLEMW2IOS23DBIOMACV2X5VBVCWMV11S1H3&client_secret=NB1WGADMKQ2NPNHBF51BZL3GBDGLSQ15Z1NDPCX25JIHQMOB&v=20180605&ll=43.6795626,-79.37752940000001&radius=500&limit=100'

In [144]:
results = requests.get(url).json()  # the results are long so I didn't show them

In [145]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [146]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(40)

name  categories        lat        lng
0         Rosedale Park  Playground  43.682328 -79.378934
1          Whitney Park        Park  43.682036 -79.373788
2  Alex Murray Parkette        Park  43.678300 -79.382773
3        Milkman's Lane       Trail  43.676352 -79.373842

In [147]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


In [148]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [149]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie


In [150]:
print(toronto_venues.shape)
toronto_venues.head()

(1274, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                    Rosedale              43.679563              -79.377529   
1                    Rosedale              43.679563              -79.377529   
2                    Rosedale              43.679563              -79.377529   
3                    Rosedale              43.679563              -79.377529   
4  Cabbagetown,St. James Town              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

In [81]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            55   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     13   
Cabbagetown,St. James Town                                             44   
Central Bay Street                                                     81   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               15   
Church and Wellesley                                                   86   
Commerce Court,Victoria Hotel                                         100   
Design Exchange,Toronto Dominion Centre                               100   
First Canadian Place,Underground city                                 100   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
Harbourfront,Regent Park                                               50   
Rosedale                                                                4   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        92   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      13   
Cabbagetown,St. James Town                                              44   
Central Bay Street                                                      81   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    86   
Commerce Court,Victoria Hotel                                          100   
Design Exchange,Toronto Dominion Centre                                100   
First Canadian Place,Underground city                                  100   
Harbord,University of Toronto                                           34   
Harbourfront East,Toronto Islands,Union Station                        100   
Harbourfront,Regent Park                                                50   
Rosedale                                                                 4   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         92   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                100             100   
Berczy Park                                            55              55   
CN Tower,Bathurst Quay,Island airport,Harbourfr...     13              13   
Cabbagetown,St. James Town                             44              44   
Central Bay Street                                     81              81   
Chinatown,Grange Park,Kensington Market               100             100   
Christie                                               15              15   
Church and Wellesley                                   86              86   
Commerce Court,Victoria Hotel                         100             100   
Design Exchange,Toronto Dominion Centre               100             100   
First Canadian Place,Under

In [84]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  \
0                    0             0         0                 0            0   
1                    0             0         0                 0            0   
2                    0             0         0                 0            0   
3                    0             0         0                 0            0   
4                    0             0         0                 0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  BBQ Joint  Baby Store  \
0           0                    0                 0          0           0   
1           0                    0                 0          0           0   
2           0                    0                 0          0           0   
3           0                    0                 0          0           0   
4           0                    0                 0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  \
0         0           0                   0       0              0          0   
1         0           0                   0       0              0          0   
2         0           0                   0       0              0          0   
3         0           0                   0       0              0          0   
4         0           0                   0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Butcher  Café  Caribbean Restaurant  \
0         0             0              0        0     0                     0   
1         0             0              0        0     0                     0   
2         0             0              0        0     0                     0   
3         0             0              0        0     0                     0   
4         0             0              0        0     0                     0   

   Cheese Shop  Chinese Restaurant  Chocolate Shop  Church  Clothing Store  \
0            0                   0               0       0               0   
1    

In [83]:
toronto_onehot.shape

(1274, 202)

In [87]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                         Berczy Park     0.000000   
2   CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
3                          Cabbagetown,St. James Town     0.000000   
4                                  Central Bay Street     0.012346   
5             Chinatown,Grange Park,Kensington Market     0.000000   
6                                            Christie     0.000000   
7                                Church and Wellesley     0.011628   
8                       Commerce Court,Victoria Hotel     0.000000   
9             Design Exchange,Toronto Dominion Centre     0.000000   
10              First Canadian Place,Underground city     0.000000   
11                      Harbord,University of Toronto     0.000000   
12    Harbourfront East,Toronto Islands,Union Station     0.000000   
13                           Harbourfront,Regent Park     0.020000   
14                                           Rosedale     0.000000   
15                            Ryerson,Garden District     0.000000   
16                                     St. James Town     0.000000   
17                    Stn A PO Boxes 25 The Esplanade     0.000000   

    Accessories Store  Adult Boutique  Afghan Restaurant   Airport  \
0                0.01        0.000000           0.000000  0.000000   
1                0.00        0.000000           0.000000  0.000000   
2                0.00        0.000000           0.000000  0.076923   
3                0.00        0.000000           0.000000  0.000000   
4                0.00        0.000000           0.000000  0.000000   
5                0.00        0.000000           0.000000  0.000000   
6                0.00        0.000000           0.000000  0.000000   
7                0.00        0.011628           0.011628  0.000000   
8                0.00        0.000000           0.000000  0.000000   
9                0.00        0.000000           0.000000  0.000000   
10               0.00        0.000000           0.000000  0.000000   
11               0.00        0.000000           0.000000  0.000000   
12               0.00        0.000000           0.000000  0.000000   
13               0.00        0.000000           0.000000  0.000000   
14               0.00        0.000000           0.000000  0.000000   
15               0.00        0.000000           0.000000  0.000000   
16               0.01        0.000000           0.000000  0.000000   
17               0.00        0.000000           0.000000  0.000000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0             0.000000        0.000000         0.000000          0.000000   
1             0.000000        0.000000         0.000000          0.000000   
2             0.076923        0.153846         0.153846          0.153846   
3             0.000000        0.000000         0.000000          0.000000   
4             0.000000        0.000000         0.000000          0.000000   
5             0.000000        0.000000         0.000000          0.000000   
6             0.000000        0.000000         0.000000          0.000000   
7             0.000000        0.000000         0.000000          0.000000   
8             0.000000        0.000000         0.000000          0.000000   
9             0.000000        0.000000         0.000000          0.000000   
10            0.000000        0.000000         0.000000          0.000000   
11            0.000000        0.000000         0.000000          0.000000   
12            0.000000        0.000000         0.000000          0.000000   
13            0.000000        0.000000         0.000000          0.000000   
14            0.000000        0.000000         0.000000          0.000000   
15            0.000000        0.000000         0.000000          0.000000   
16            0.000000        0.000000         0.000000          0.000

In [89]:
toronto_grouped.shape

(18, 202)

In [91]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.06
2      Thai Restaurant  0.04
3           Steakhouse  0.04
4  American Restaurant  0.04


----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2    Restaurant  0.04
3          Café  0.04
4   Cheese Shop  0.04


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.15
1  Airport Terminal  0.15
2    Airport Lounge  0.15
3             Plane  0.08
4  Sculpture Garden  0.08


----Cabbagetown,St. James Town----
                venue  freq
0         Coffee Shop  0.09
1          Restaurant  0.07
2                Café  0.05
3  Italian Restaurant  0.05
4                 Pub  0.05


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.15
1                 Café  0.06
2   Italian Restaurant  0.05
3  Japanese Restaurant  

In [92]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [96]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2   CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
3                          Cabbagetown,St. James Town           Coffee Shop   
4                                  Central Bay Street           Coffee Shop   
5             Chinatown,Grange Park,Kensington Market                  Café   
6                                            Christie         Grocery Store   
7                                Church and Wellesley   Japanese Restaurant   
8                       Commerce Court,Victoria Hotel           Coffee Shop   
9             Design Exchange,Toronto Dominion Centre           Coffee Shop   
10              First Canadian Place,Underground city           Coffee Shop   
11                      Harbord,University of Toronto                  Café   
12    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
13                           Harbourfront,Regent Park           Coffee Shop   
14                                           Rosedale                  Park   
15                            Ryerson,Garden District           Coffee Shop   
16                                     St. James Town           Coffee Shop   
17                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Café            American Restaurant            Steakhouse   
1           Cocktail Bar                       Beer Bar                  Café   
2       Airport Terminal                Airport Service       Harbor / Marina   
3             Restaurant                            Pub     Indian Restaurant   
4                   Café             Italian Restaurant        Ice Cream Shop   
5                    Bar  Vegetarian / Vegan Restaurant    Chinese Restaurant   
6                   Café                           Park            Baby Store   
7            Coffee Shop                        Gay Bar      Sushi Restaurant   
8                   Café                          Hotel            Restaurant   
9                  Hotel                           Café   American Restaurant   
10                  Café                          Hotel            Restaurant   
11             Bookstore                    Coffee Shop   Japanese Restaurant   
12                 Hotel                           Café           Pizza Place   
13                  Park                         Bakery                  Café   
14            Playground                          Trail         Women's Store   
15        Clothing Store                 Cosmetics Shop                  Café   
16                  Café                     Restaurant        Clothing Store   
17                  Café                     Restaurant              Beer Bar   

   5th Most Common Venue  6th Most Common Venue      7th Most Common Venue  \
0        Thai Restaurant                    Bar                        Gym   
1            Cheese Shop             Steakhouse                 Restaurant   
2       Sculpture Garden               Boutique                      Plane   
3     Italian Restaurant                 Bakery                Pizza Place   
4                    Bar        Bubble Tea Shop               Burger Joint   
5     Mexican Restaurant  Vietnamese Restaurant                     Bakery   
6     Italian Restaurant                  Diner                  Nightclub   
7             Restaurant           Burger Joint            Bubble Tea Shop   
8    American Restaurant     Seafood Restaurant         Italian Restaurant   
9             Restaurant          Deli / Bodega                  Gastropub   
10            Steakhouse    American Restaurant              Deli / Bodega   
11                   Bar                 Bakery                

## 4. Cluster Neighborhoods

In [99]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 3, 0, 4, 0, 3, 3], dtype=int32)

In [101]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postcode   Latitude  Longitude           Borough  \
0      M4W  43.679563 -79.377529  Downtown Toronto   
1      M4X  43.667967 -79.367675  Downtown Toronto   
2      M4Y  43.665860 -79.383160  Downtown Toronto   
3      M5A  43.654260 -79.360636  Downtown Toronto   
4      M5B  43.657162 -79.378937  Downtown Toronto   

                 Neighborhood  Cluster Labels 1st Most Common Venue  \
0                    Rosedale               0                  Park   
1  Cabbagetown,St. James Town               0           Coffee Shop   
2        Church and Wellesley               2   Japanese Restaurant   
3    Harbourfront,Regent Park               0           Coffee Shop   
4     Ryerson,Garden District               3           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail         Women's Store   
1            Restaurant                   Pub     Indian Restaurant   
2           Coffee Shop               Gay Bar      Sushi Restaurant   
3                  Park                Bakery                  Café   
4        Clothing Store        Cosmetics Shop                  Café   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0      Department Store     Electronics Store        Dumpling Restaurant   
1    Italian Restaurant                Bakery                Pizza Place   
2            Restaurant          Burger Joint            Bubble Tea Shop   
3        Breakfast Spot                   Pub         Mexican Restaurant   
4                   Bar   Japanese Restaurant  Middle Eastern Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop      Doner Restaurant                Dog Run  
1                  Café  Caribbean Restaurant         Breakfast Spot  
2                   Pub           Men's Store                   Café  
3          Dessert Shop            Restaurant                Brewery  
4            Restaurant                 Diner            Pizza Place

In [102]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

In [105]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Latitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   43.679563               0                  Park            Playground   
1   43.667967               0           Coffee Shop            Restaurant   
3   43.654260               0           Coffee Shop                  Park   
5   43.651494               0           Coffee Shop                  Café   
7   43.657952               0           Coffee Shop                  Café   
11  43.648198               0           Coffee Shop                  Café   
15  43.646435               0           Coffee Shop                  Café   
16  43.648429               0           Coffee Shop                  Café   
17  43.669542               0         Grocery Store                  Café   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Trail         Women's Store      Department Store   
1                    Pub     Indian Restaurant    Italian Restaurant   
3                 Bakery                  Café        Breakfast Spot   
5             Restaurant        Clothing Store                 Hotel   
7     Italian Restaurant        Ice Cream Shop                   Bar   
11                 Hotel            Restaurant   American Restaurant   
15            Restaurant              Beer Bar    Seafood Restaurant   
16                 Hotel            Restaurant            Steakhouse   
17                  Park            Baby Store    Italian Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Electronics Store   Dumpling Restaurant            Donut Shop   
1                 Bakery           Pizza Place                  Café   
3                    Pub    Mexican Restaurant          Dessert Shop   
5           Cocktail Bar   Japanese Restaurant    Italian Restaurant   
7        Bubble Tea Shop          Burger Joint   Japanese Restaurant   
11    Seafood Restaurant    Italian Restaurant             Gastropub   
15          Cocktail Bar                 Hotel   Japanese Restaurant   
16   American Restaurant         Deli / Bodega                   Bar   
17                 Diner             Nightclub     Convenience Store   

   9th Most Common Venue 10th Most Common Venue  
0       Doner Restaurant                Dog Run  
1   Caribbean Restaurant         Breakfast Spot  
3             Restaurant                Brewery  
5         Cosmetics Shop              Gastropub  
7      Indian Restaurant     Falafel Restaurant  
11         Deli / Bodega             Steakhouse  
15           Cheese Shop                 Bakery  
16             Gastropub                    Gym  
17            Restaurant            Coffee Shop